Remove contemplation of invariant violations from delta_fifo.go

MikeSpreitzer opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug

/kind cleanup

> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Some comments and code incorrectly contemplated violating the
invariant that a keys is in `f.items` if and only if it is in
`f.queue`.  Changed to not contemplate invariant violations,
and panic if one is stumbled upon.

Also fixed up some comment wording, some of which was discussed late in #91435 

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #91709 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MikeSpreitzer committed:
Remove contemplation of invariant violations from delta_fifo.go

Some comments and code incorrectly contemplated violating the
invariant that a keys is in `f.items` if and only if it is in
`f.queue`.

Also fixed up some comment wording.

MikeSpreitzer commented:
@kubernetes/sig-api-machinery-bugs

MikeSpreitzer commented:
/assign @lavalamp 
@jqmichael

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -161,12 +161,14 @@ type DeltaFIFO struct {
 	lock sync.RWMutex
 	cond sync.Cond
 
+	// We maintain the properties that:
+	// - a key is in `items` if and only if that key is in `queue`;
+	// - a given key appears in at most one position in `queue`;
+	// - each Deltas in `items` has at least one Delta.
```
What's your thoughts of moving this block below the definition of `queue` to minimize forward references?

MikeSpreitzer replied:
Reorganized comments to remove forward references.

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -161,12 +161,14 @@ type DeltaFIFO struct {
 
+	// We maintain the properties that:
+	// - a key is in `items` if and only if that key is in `queue`;
+	// - a given key appears in at most one position in `queue`;
```
What's your thoughts of simplifying this statement to `There's no duplicates in queue`. 
When I see wordings like `at most`, I always think of something QoS related, i.e. "at most once", "at least once", "exactly once" etc.

MikeSpreitzer replied:
Reworded.

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -389,10 +391,10 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
 		// when given a non-empty list (as it is here).
-		// But if somehow it ever does return an empty list, then
-		// We need to remove this from our map (extra items in the queue are
-		// ignored if they are not in the map).
-		delete(f.items, id)
+		// But if somehow it ever did return an empty list, then
+		// it would not be legitimate to simply remove the entry for `id`
+		// from `f.items` without also ensuring that `id` is absent from `f.queue`.
```
I think it's sufficient to say `This never happens, because ...` and then add a `panic()`. At least that would be more intuitive for me as a reader. 

What do you think?

MikeSpreitzer replied:
I think you are suggesting removing these lines:
```
		// But if somehow it ever did return an empty list, then
		// it would not be legitimate to simply remove the entry for `id`
		// from `f.items` without also ensuring that `id` is absent from `f.queue`.
```

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -476,37 +478,35 @@ func (f *DeltaFIFO) IsClosed() bool {
+		f.initialPopulationCount--
 	}
+	item, ok := f.items[id]
+	if !ok {
+		// This should never happen
+		panic(id)
+	}
```
Can we be more specific in the error message to help troubleshooting in case it does occur (maybe due to a bug in the future)?

MikeSpreitzer replied:
Revised to pass an error message to `panic`.

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -476,37 +478,35 @@ func (f *DeltaFIFO) IsClosed() bool {
+	delete(f.items, id)
+	err := process(item)
+	if e, ok := err.(ErrRequeue); ok {
+		f.addIfNotPresent(id, item)
```
Out of scope of this PR, but I've always been wondering, what is the point of calling `addIfNotPresent` in deltaFIFO. The entire `pop()` operation was invoked under lock (acquired after `cond.wait()`), and we deleted the key from both `queue` and `items`, and there's no way the key would still exist after `process()`, right?

jqmichael commented:
/lgtm 

@lavalamp

jqmichael commented:
/retest

MikeSpreitzer commented:
/test pull-kubernetes-integration

MikeSpreitzer commented:
/assign @wojtek-t 
Can you please take a look at this?

MikeSpreitzer commented:
/unassign @lavalamp

wojtek-t commented:
Looking.

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -389,10 +390,7 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
-		// We need to remove this from our map (extra items in the queue are
-		// ignored if they are not in the map).
-		delete(f.items, id)
+		panic(fmt.Sprintf("Impossible dedupDeltas for id=%q: old items=%#+v, obj=%#+v", id, f.items[id], obj))
```
Let's not panic - if because of some reason we have a bug and this would happen, we would never recover from crashloop in such case.
Instead let's simply return an error here.

MikeSpreitzer replied:
I have had others advise me that the panic is actually the preferred response.  I think the theory goes something like the following.  The question is how can code react when it detects data corruption that must be due to a bug that presumably bites in rare circumstances (if it bit commonly then the code would not have made it past CI, and there is no response that will save the day).  If the in-memory data is now corrupted then keeping the process alive with the corrupted data is going to run into more trouble than restarting the process and letting it run successfully until such time, if any, when the bug bites again.

wojtek-t replied:
I don't think I fully agree.

In cases like this, data-corruption basically happens due to some bug in the code. We are all just humans - while we believe it should never happen, noone knows if there isn't any subtle bug still around.
So if it happened once, it can also be repeatable and we can enter the crashloop.

So while in test environment it would be desired to panic (and even enter crashloop), it production, I would like to get the situation where because of bug in the code I'm not able to exit this crashloop (with the current state of etcd) and the only way to fix my production cluster is to fix the bug and rebuild it.

So I'm really opposed to switching to panics here.

@liggitt @lavalamp - for thoughts

jqmichael replied:
Failing fast vs tolerance against errors is always controversial.  :)

I guess the decision is case by case. It depends on how critical maintaining such data invariant is such that we are willing to shut down the entire system if we end up breaking that invariant.

MikeSpreitzer replied:
It seems to me that the only good answer is to write code to repair the data structure.  In this case I think that would be fairly straightforward, no missing critical information to guess at.

MikeSpreitzer replied:
Latest revision repairs instead of panicing or returning an error.

wojtek-t commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -476,37 +474,35 @@ func (f *DeltaFIFO) IsClosed() bool {
+	item, ok := f.items[id]
+	if !ok {
+		// This should never happen
+		panic(fmt.Sprintf("Inconceivable! %q was in f.queue but not f.items!", id))
```
Same here - let's not introduce panics and simply return an error.

wojtek-t reviewed (commented):
Two small comments.

MikeSpreitzer committed:
Repair instead of panic when data corruption detected in DeltaFIFO

MikeSpreitzer commented:
/retest

jqmichael commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -650,6 +656,35 @@ func (f *DeltaFIFO) syncKeyLocked(key string) error {
 	return nil
 }
 
+// A partial repair function.  Removes keys from f.queue that
+// correspond to nothing in f.items.  Removes keys from both when the
+// entry in f.items has zero length.  Ensures each key appears at most
+// once in queue.  Must be called with the lock held.
+func (f *DeltaFIFO) filterQueueLocked() {
```
Just to share my two cents.

Personally, I think adding 29 lines of code (3.7% of the entire file) in a place which is supposed to be dead code is too much maintenance overhead. Any new reader of this class will have to try to understand what this function does. 

We often say we spent 80% of efforts in places which only occur 20% of the time, which is a big waste. I think this is one of such places. We have to draw a balance between 100% of code correctness and the increased maintenance overhead. 

I would rather not to pay the maintenance cost today for some bug which doesn't yet exist. Therefore, my vote for now is to simply return an error if dead code does get triggered due to a bug. We can fix the bug after with sufficient logging. 

Thoughts?

MikeSpreitzer replied:
I don't like a lot of dead code either.  So I revised again, this time returning an error when in a tight spot.  This does not avoid breaking the data structure, and introduces more possible error returns (so more previously-unexercised logic in the callers (all the clients of delta_fifo) may get exercised in that case).

MikeSpreitzer committed:
Replaced repair with returning error, in delta_fifo.go

When dedupDeltas does the impossible and the key is already queued,
return an error rather than maintain the data structure invariants.

jqmichael commented:
/retest 

@wojtek-t

wojtek-t commented:
/lgtm
/approve

Thanks!

justaugustus commented:
/test pull-kubernetes-e2e-kind-ipv6
